packages/hebao_v1/contracts/modules/dapps/base/GenericDAppModule.sol

```diff
@@ -17,111 +17,163 @@
-import "../../../lib/Claimable.sol";
-import "../../../thirdparty/BytesUtil.sol";
-
 import "../../security/SecurityModule.sol";
 
+import "../../security/GuardianUtils.sol";
+
 
 /// @title GenericDAppModule
 /// @dev GenericDAppModule allows wallet owners to transact directly or through meta
-///      transactions on any whitelisted dApps. The transaction data must be appended
-///      with the wallet address and then the dApp's address.
-contract GenericDAppModule is Claimable, AddressSet, SecurityModule
+///      transactions on any whitelisted dApps.
+contract GenericDAppModule is SecurityModule
 {
-    using BytesUtil for bytes;
+    enum SecuritySetting
+    {
+        Owner,
+        Majority
+    }
 
-    bytes32 internal constant DAPPS = keccak256("__DAPP__");
+    mapping (address => SecuritySetting) public securitySettings;
+    mapping (address => mapping(address => bool)) public approvedDApps;
 
-    event DAppEnabled(address indexed dapp, bool enabled);
+    modifier onlyApprovedDapp(
+        address          wallet,
+        address          dapp
+        )
+    {
+        require(isDAppApproved(wallet, dapp), "DAPP_NOT_APPROVED");
+        _;
+    }
+
+    modifier onlyWhenAuthorized(
+        address           wallet,
+        address[] memory  signers
+        )
+    {
+        SecuritySetting securitySetting = securitySettings[wallet];
+        if (securitySetting == SecuritySetting.Owner) {
+            address walletOwner = Wallet(wallet).owner();
+            require(
+                msg.sender == walletOwner ||
+                signers.length == 1 && signers[0] == walletOwner,
+                "UNAUTHORIZED"
+            );
+        } else if(securitySetting == SecuritySetting.Majority) {
+            GuardianUtils.requireSufficientSigners(
+                securityStore,
+                wallet,
+                signers,
+                GuardianUtils.SigRequirement.OwnerRequired
+            );
+        } else {
+            revert("UNKNOWN_SECURITY");
+        }
+        _;
+    }
 
     constructor(Controller _controller)
         public
-        Claimable()
         SecurityModule(_controller)
     {
     }
 
-    function enableDApp(address dapp)
+    function setSecurity(
+        address            wallet,
+        address[] calldata signers,
+        uint8              _securitySetting
+        )
         external
-        onlyOwner
+        nonReentrant
+        onlyFromMetaTxOrWalletOwner(wallet)
+        onlyWhenWalletUnlocked(wallet)
+        onlyWhenAuthorized(wallet, signers)
     {
-        require(
-            !controller.moduleRegistry().isModuleRegistered(dapp),
-            "MODULE_NOT_SUPPORTED"
-        );
-        require(
-            !controller.walletRegistry().isWalletRegistered(dapp),
-            "WALLET_NOT_SUPPORTED"
-        );
-        addAddressToSet(DAPPS, dapp, true);
-        emit DAppEnabled(dapp, true);
+        securitySettings[wallet] = SecuritySetting(_securitySetting);
```

Comment:
when will an owner needs to use `setSecurity` and `setDAppApproved ` ?

Reply:
If the default remains `SecuritySetting.Owner` than never. But the UI could expose an extra security setting to change this optionally for users that want extra security for dapps.